js/src/vector/float.ts

```diff
@@ -18,23 +18,49 @@
 import { Data } from '../data';
 import { Vector } from '../vector';
 import { BaseVector } from './base';
+import { Vector as V } from '../interfaces';
 import { Float, Float16, Float32, Float64 } from '../type';
+import { toFloat16Array, toFloat32Array, toFloat64Array } from '../util/buffer';
 
 export class FloatVector<T extends Float = Float> extends BaseVector<T> {
-    public static from<T extends Float16>(data: T['TArray']): Float16Vector;
-    public static from<T extends Float32>(data: T['TArray']): Float32Vector;
-    public static from<T extends Float64>(data: T['TArray']): Float64Vector;
+
+    public static from(this: typeof FloatVector, data: Float16['TArray']): Float16Vector;
+    public static from(this: typeof FloatVector, data: Float32['TArray']): Float32Vector;
+    public static from(this: typeof FloatVector, data: Float64['TArray']): Float64Vector;
+    public static from<T extends Float>(this: typeof FloatVector, data: T['TArray']): V<T>;
+
+    public static from(this: typeof Float16Vector, data: Float16['TArray'] | Iterable<number>): Float16Vector;
+    public static from(this: typeof Float32Vector, data: Float32['TArray'] | Iterable<number>): Float32Vector;
+    public static from(this: typeof Float64Vector, data: Float64['TArray'] | Iterable<number>): Float64Vector;
```

Comment:
I don't really follow this syntax, with the `this` argument, could you clarify what's going on here?

I assume it somehow makes it so that both `FloatVector.from(new Float32Array())` and `Float32Vector.from([1,2,3])` yield a `Float32Vector` but I don't really understand how.

Reply:
@TheNeuralBit yep! In plain English it'd be something like, "If `this` == Float16Vector constructor, accept data as either Uint16Array or any Iterable<number>, and always return an instance of Float16Vector." It's there to ensure that `Float16Vector.from([1, 2, 3]) instanceof Float16Vector` is always true at compile _and_ run-times.

TypeScript doesn't grant much latitude when it comes to subclasses overriding inherited static functions, so we have to hoist the narrower type definitions up to the base. You can also [see here](https://github.com/trxcllnt/arrow/pull/8/files/2557277087f8f86b96715fe1bbe7ec88aa443b71#diff-a92b06b2ed35319c47a86b2b08d0947cR59) how I figured out how to check this case in the unit tests.